### ISSUE TITLE
Single table design and partition/sort keys

### ISSUE THREAD

**Issue by MickL** [dynamoosejs/dynamoose#768]
This is more like a question than an issue.

The AWS DynamoDB documentation says a good designed application uses only one table. Also I found a very good introduction into data modeling in [this video](https://youtu.be/DIQVJqiSUkE) from AWS re:Invent 2019.

The key concepts of DynamoDB modeling seems to be to set the right partition keys and sort keys to build up the relationships and queries. However in Dynamoose it seems like every model creates a new table and i didnt find any informations about setting partition/sort keys for each model and making the partition/sort keys out multiple attributes comined. 

So my question is: Compared to modeling like in the video isnt Dynamoose implementing bad practices? How is it working compared to the techniques provided in the video? Escpecially when it comes to performance and scalability. Or is Dynamoose doing all that stuff but abstracting it so the user(programmer) does not have to worry about it?

**Comment by fishcharlie** [comment]
@MickL Extremely complex topic here. Apologizes in advance for the massive message.

---

Let me first address some points that you brought up originally.

> However in Dynamoose it seems like every model creates a new table

Correct, 1 `Model` in Dynamoose = 1 `Table` in DynamoDB

> i didnt find any informations about setting partition/sort keys for each model

Looks like an oversight in v2. In v1 there is an option here. Will work on fixing this. By default the hash key is set to the first item in the schema. In v1 you could change that behavior and set a specific one, in v2 looks like I haven't built that functionality yet. Will work on that.

> Compared to modeling like in the video isnt Dynamoose implementing bad practices?

Although I was at re:Invent 2019, I didn't get a chance to attend that session. It's on my todo list to watch that video, but I haven't gotten the chance yet. If I had to guess, it discusses that since it's NoSQL having one table is better than having relationships across multiple tables since it's not a relationship based database.

I would **not** classify it as *implementing bad practices* tho. Dynamoose was inspired heavily by Mongoose. There are some areas where we are kinda behind in Mongoose. But I think in general, if you make the argument that Dynamoose is implementing bad practices, then I'd make the argument that Mongoose is as well. I just don't think that is a fair categorization of what is going on here.

At the end of the day, Dynamoose is a tool. It's the developers responsibility to use best practices for database modeling. Dynamoose isn't a magic wand that replaces the need for strong database modeling practices, and good database design decisions. That isn't the purpose of Dynamoose or any other similar library. Nor will it ever be.

That being said, Dynamoose should take responsibility for implementing features, functionality, and encouraging users to implement the best database modeling and design practices possible (will get into this more later).

> Or is Dynamoose doing all that stuff but abstracting it so the user(programmer) does not have to worry about it?

Highly doubtful. But again, haven't watched that video.

I will prioritize watching that session a bit more and get back to you with my thoughts.

---

In the meantime. Let me just tell you about some broad, higher level initiates I've been working on for Dynamoose that I believe do effect what you are talking about.

First of all, Dynamoose is going through a complete rewrite (v2, #709). I believe this will set us up to do some amazing things in the future in regards to what you are talking about. I believe the foundation of Dynamoose is stronger than it's ever been.

One of the things that has been discussed in the past is support for nested schemas/models (#301). More nested schemas I think makes more sense tho. But for example, where inside of one schema you can reference another schema and have that behavior work as expected.

In the [breaking changes for v2](https://github.com/dynamoosejs/dynamoose/blob/master/BREAKING_CHANGES.md), one of the items is `expires has been moved from the Schema settings to the Model settings`. This is just one of many examples of where throughout the v2 rewrite I've focused on decoupling the Schema and Model's from one another. The one thing I dislike about this, is it does modify the Schema by adding an attribute to the Schema for the expires/ttl property. Which is really not fun at all, and I don't think uses a good philosophy here. But I'm struggling to think about how to redesign this in a better way (any input would be great).

Another example of this (that I need to add to the breaking changes list), is that the `throughput` setting has been moved from the Schema to the Model settings. Again for the same reason of decoupling the Schema and Model from one another and truly thinking about what makes sense for each. Model has the underlying concept of a Table, so knowing which properties go to it should be obvious (ex. throughput, since it's a Table concept).

Overall tho, there has been a lot of work to ensure that Schema and Model are two completely separate concepts with very little overlap. Things that are set per table (ex. expires/ttl, throughput) should be contained within the model, not the schema.

I have to look at how this applies to Indexes a bit more. I don't think you are allowed to have Indexes in nested properties (objects, arrays) in DynamoDB. Currently that is handled within the Schema tho. Due to that it might not make sense to have that in the Schema anymore and move it to the Model.

But even what I mentioned earlier about custom hash and range key stuff, that should be part of the Model settings probably. Only one exist per Model/Table. In Dynamoose v1 it was handled within the Schema. That probably doesn't make sense. But on the other hand, if you don't set the property, it will derive it from the Schema. Which would feel slightly weird to have separation there.

Although I don't anticipate nested schema support happening in v2.0, I do see it potentially happening in a v2.1 release or shortly thereafter. I think Dynamoose v2 is incredibly well prepared for some amazing stuff in this regard. Dynamoose v2 has incredible test coverage, and fixes a lot of pesky bugs that have been lingering in the code base for a while (ex. #669). I think it also separates things out and truly has the best foundation.

---

v2 is still in beta. I'm using it in production in a few of my apps. I've found it to be stable in my use cases. Some functionality hasn't been built out yet, and there are still some issues (ex. creation and updating of tables), but overall it's incredibly stable.

Here would be my asks for you. Don't feel obligated to do any of these, but it would help me with what you are talking about, and progressing this.

1. Try to implement v2 beta, see how it goes for you. Give me feedback (pros and cons). What works, what doesn't, what needs to be improved?
2. Look through the code and documentation for v2. Let me know if anything stands out as being problematic for the foundation and structure for v2. This is the time to make breaking changes to the code/API. If there is an option that you think should be on the Schema vs the Model or visa-versa, let me know. If you think there is something we could do the make the foundation better for the future, let me know. This is really the time that I'm looking for feedback and people to dive into it. Question everything. Are we doing things the best way possible? Could things be done better? I want the foundation and code base to be **AMAZING** right now, that way for future features, it's easier to develop. Is the folder structure, and the way utility functions setup make sense? Is the code DRY enough? etc.
3. Consider joining the Slack group (link in the README). I'd be happy to discuss this more there if you want more of a chat experience. But I'm also fine with just replying here. As you might have noticed, this is a LOT tho. Might be easier to continue over a chat platform. Totally up to you tho.
4. Finally, I'd **love** a bit more information on your perspective on what concrete steps Dynamoose can take to get more inline with those suggestions from AWS. What features, functionality, changes can we make to improve?

I will try to prioritize watching that session and take some notes on what I think could be done for that 4th item as well. But I'd also love your perspective.

If you want to help further but only have limited time, for sure I'd say the 4th item is the most important. Tho the 2nd item I haven't gotten much feedback on, and is something I really wish I had more input on.

---

If you got this far in reading this, **THANK YOU 🎉**. I know it was a lot, but I feel like your comments and further discussion has the potential to make Dynamoose better. So I thank you for that, and I look forward to discussing further.

**Comment by fishcharlie** [comment]
@MickL After sending my previous message I watched that re:Invent session video.

Here is what I came up with what Dynamoose could maybe add and contribute here. Each of the items below represents a new feature Dynamoose could potentially add to make this process easier for developers.

| Video Timestamp Link | Video Notes | Description | Potential Problems | Questions |
| --- | --- | --- | --- | --- |
| https://youtu.be/DIQVJqiSUkE?t=1441 | Having an Address entity be part of the existing User entity, but also have some level of separation there | This is **exactly** what I was talking about earlier with nested Schema support. 100% on my radar and something that will hopefully be coming soon. The more support I get with v2 and the things that need to happen before that, the quicker that will happen. | Unaware of any potential problems at this stage. v2 seems to be well prepared for this, but still need to investigate further. |   |
| https://youtu.be/DIQVJqiSUkE?t=1594 |   | For one Table/Model you can have multiple top level/root Schema's. Maybe by allowing you to pass in an array of Schema's into the Model init?? | Will need to investigate further into how Dynamoose can detect which Schema is being used, and handle appropriately. Maybe the developer will have to specify which Schema to use?? Not sure exactly how this will look on the implementation side, but seems like it could be doable. |  |
| https://youtu.be/DIQVJqiSUkE?t=2121 | Slide talks about combining attributes into single attribute | Without much thought this seems like it could be possible to implement into Dynamoose and make native without effort from the dev. | One pitfall we'll have to consider is what happens when you do an `updateItem`. Doing a `putItem` is easy in this case since it overwrites or creates the entire document. But when you do a `updateItem` you don't necessarily have access to all the attributes or the final document at time of saving (especially if you do an `$ADD` operation or aren't updating a specific property that is part of that combined attribute). |  |
| https://youtu.be/DIQVJqiSUkE?t=2325 | If Attribute A is a certain value, Attribute B is set to a hash or something. If Attribute A is anything other than that certain value, Attribute B is set to undefined. Attribute B is also an index. | This seems like it could be implemented into Dynamoose. | The only problem is figuring out what to make the value of Attribute B be. Maybe there is room to have an function on the Schema attribute settings that we call and pass in the existing Document or something and have the function return the hash or whatever you want the value to be. | I'm not sure why exactly you couldn't have just had a index on the Status attribute and queried on that for where `Status = PLACED`. I'm assuming this is because you don't want the index size to get too large? |

Those are the things that stood out to me in the video that Dynamoose doesn't handle as efficiently as it could.

Keep in mind, **everything** that I mentioned above, is possible with Dynamoose today. There is no functionality that I heard or saw that isn't possible with Dynamoose today. It just might not be as good as it could be.

One example of this would be for the first item above, you can use the Object type in Dynamoose to specify information and a schema for that object (tho again, it would be cool to be able to just pass a Schema directly in there). Another example would be having multiple entities for one Model/Table, you can do this by setting the Schema attributes and just making them not required. Again, not as cool as if you could pass in an array of Schema options into the Model init, but still possible to do this today.

---

Hopefully that helps you a little bit. Let me know your thoughts and if I missed anything here. Doubtful any of this will be implemented into Dynamoose v2.0. But likely will come in v2.1 or sometime there after. The more support I can get in ensuring a solid foundation for Dynamoose v2 and the more testing I can get for it, the better. If you can provide in helping with Dynamoose v2. The sooner we ship v2 the sooner we will be able to turn our focus to the items mentioned above. But this is open source, so the project is only as good as the community and the time people put into it. So it's all a tradeoff and the more support the project can get, the more likely the things above will be implemented (and implemented quickly).

**Comment by MickL** [comment]
Woah i didnt expect such a detailed response! Thank you so much!

I would like to contribute and will definitely give you feedback but yes time is a limiting factor. Also I am kind of new to DynamoDB. I used MongoDB with Mongoose before, watched the video and was then totally confused seeing Dynamoose. Thats why I asked the questions. In that regard I can give you limited feedback only. 

First thing about the v2 documentation I noticed is that the files are just there. It would be nice to have a website where the md files automatically generate a documentation so at least the topics have an order. Having not much knowledge about the buzzwords it is hard to find an order for me.

I have one more question: What do you think about `dynamo-types`? It uses decorators which i really like. But anyhow I was wondering if they do things differently and maybe even if it makes sense to group up with them? Maybe they solved already some of the pitfalls you ran into? Or maybe there is a good reason to have both or to favor dynamoose. I was just about to use their project in favor of dynamoose because of the decorators and a lot of stars means it might be largely tested. I didnt know that you put so much effort and knowledge into this project, too.

**Comment by fishcharlie** [comment]
> Woah i didnt expect such a detailed response! Thank you so much!

Of course! Again, I apologize for the long messages. Your question could have been taken in so many different directions, which is awesome, but just wanted to make sure I covered as many angles as possible.

> I would like to contribute and will definitely give you feedback but yes time is a limiting factor.

Totally understand. If/as you use Dynamoose more you can hopefully give more feedback! 😃 

> First thing about the v2 documentation I noticed is that the files are just there. It would be nice to have a website where the md files automatically generate a documentation so at least the topics have an order.

**Totally agree** with this. It's on my todo list. Likely won't get done until the final release since I'm more focused on stability and ensuring we have great test coverage and stuff. Dynamoose v1 had around 300 tests with roughly 80% code coverage I think? Dynamoose v2 now has over 1000 tests with 100% code coverage. Although I know code coverage isn't the only factor, it does play a role in stability I believe. So really been focusing on that more so than a website. But documentation I think is much better than v1 as well.

Will for sure get the website updated before v2 is finalized.

> What do you think about dynamo-types?

Do you mean https://www.npmjs.com/package/dynamo-types? That is the closest match I found, but you mentioned that it has `a lot of stars means it might be largely tested`, but the [GitHub repo](https://github.com/balmbees/dynamo-types) only has 123 stars. So I'm not quite sure what you are referring to.

I can reply more about your questions about that once we are on the same page 😃 

> I didnt know that you put so much effort and knowledge into this project, too.

For years now 😆. It's by far the open source project I contribute to most. Mainly because I use it in almost every one of my projects. I can't afford for this project to be bad or subpar because I use it so often. I truly want it to be the best it can possibly be.

**Comment by MickL** [comment]
Yes it is this project: https://github.com/balmbees/dynamo-types

Would love to hear your opinion on this one compared to Dynamoose and maybe you find solutions or other ways to so things in the source code?

**Comment by fishcharlie** [comment]
> Would love to hear your opinion on this one compared to Dynamoose

I know I'm biased. But I'd choose Dynamoose over that everyday. They do have their similarities and differences. Kinda similar to the tons of web frameworks there are for Node.js. Each has a slightly different style.

I'd pick Dynamoose any day tho because of it's similarity to Mongoose, which is a very well known and established player for MongoDB. I also question the functionality and features of `dynamo-types`. For example, when doing a `batchPut`, why do you need to declare a `writer` static property on the class? That feels one step removed from what it should be.

Also I find their documentation to be extremely lacking. Dynamoose v1's documentation I feel like was pretty poor. It think it's improved a lot for v2. But I have a lot of questions regarding the documentation of `dynamo-types`.

One example of this is `await card.save({ condition: { title: "Title" } });` in the README. My biggest question here is why do you need to say `condition`? What other settings can you modify for save? Is there options to change the `ReturnValues` property that gets send with the request? Are you able to do more complex conditions like only saving if the existing key doesn't exist (overwriting vs non overwriting)?

Examples in documentation is **great**. But having examples is not a replacement for explaining the options, settings, and APIs a developer can use. I feel like `dynamo-types` is really poor in this area. The moment you want to do anything a little bit complex, custom, or unique. It's extremely difficult to find documentation for what you want since it only has very specific examples.

Also, is it even possible to set a property as `required` in `dynamo-types`? Just one example. But things like that either they don't support it, or it's not obvious in their documentation how it works.

Finally, even tho decorators can be cool at times, I'm personally not a fan of needing them for everything. It gets to be a lot when you need to repeat the same decorator for each property. I think when looking at a Schema for Dynamoose, the important information stands out **much** better in the syntax and is easier to identify. I find it much easier to process an entire schema or understand what is going on in Dynamoose just due to the simplicity of it.

> ofcourse it gives typechecking which Mongoose or Dynamoose doesnt

Dynamoose 100% does type checking. For example.

https://github.com/dynamoosejs/dynamoose/blob/cc2c09820c3aef5d920c93ec8b5cc25dbe539cd2/test/Document.js#L1009-L1015

Now, you are correct, it doesn't do type checking through like TypeScript or anything. But it 100% does type checking in general.

In terms of TypeScript, as I said before, it's on my radar and something I want to look into more for Dynamoose. But I want to implement it the correct way. In the past any integration we had with TypeScript wasn't maintained or kept up to date at all. That to me is worse than not having TypeScript integration at all.

It totally is on my radar, and hopefully will happen soon. I think after Dynamoose v2 is released, it'll really set me up to be able to focus on diving into TypeScript on some personal projects, then do a deep dive into how it makes sense to integrate it into Dynamoose.

> I wonder if it would require less tests if everything would been written in TypeScript right away.

I have 0 regrets about the decision to use JavaScript to start. I know TypeScript has a lot of support. But the reality is transpilers add another level of complexity and are another point where things can break. Which is why I have been resistant to TypeScript in the past. Less surface area to break. If for some reason TypeScript fails to convert the code to JavaScript or the output doesn't equal the input, that is a major problem. And even if it doesn't happen often, I'd argue that when it does happen, it's a **MAJOR** problem. So there is a tradeoff there.

Part of me is still holding out hope that maybe someday tc39 will work on type support for JS. I haven't done enough research on their position on it and if that is even something that is being considered. But there would be clear benefits to that over TypeScript.

However, all of that being said, TypeScript has gained a lot of popularity. It is also getting to the point where it seems to be maintaining that popularity. I've seen a lot of early adopters jump on board with major new technologies way too early only to be burned later. Due to the fact that I use Dynamoose in a lot of my production personal projects, I don't want to take the risk of jumping on board too early. [Express](https://expressjs.com) has been popular for a long time. In major part (in my opinion) due to it's stability. There is a delicate balance between resisting change, and ensuring you are adapting to changes. No one wants their system to be completely legacy that isn't modern at all, but also, no one wants their system to break due to moving to fast and aggressively on new technologies. I think just being smart about that is important and all you can do. So I standby my decision of focusing and prioritizing testing, documentation, and stability for v2.0. TypeScript is on my todo list tho.

In terms of the original question tho, maybe a few less tests? I don't think it would have been significant at all tho.

**Comment by fishcharlie** [comment]
@MickL Sounds like I answered most of your questions. Gonna close this issue for now just to keep things organized, since it's not really actionable anymore. Feel free to keep commenting tho if you have further questions.

**Comment by MickL** [comment]
Thanks for all your explainations. I will definitely support this project by using it and giving you feedback but I might run into obstacles without the typings and type checking. I remember Mongoose to be the same. There is Typegoose to fix that but it was a pain in some ways top since it builds on the same principles of Mongoose. IMO TypeORM shows the most efficient coding-experience. 

I totally understand your decision to finish v2 properly. This is definitely the right way!

Even tho it is not related right now, i just wanted to add comments to TypeScript:

- It is a new Technology. It is out for a long long time and more and more adapting for a good reason.
- The test you showed above is exactly what would not be needed anymore. The types cant be wrong. 
- It does not add any complexity layer or whatever. It just adds types, interfaces and classes. The first two doesnt even exist when compiled. The third gets to whatever output you want.
- There are no problems with the compiler output being not what you expected. 
- Personally I will never ever go any project without it anymore. JavaScript with no typings is a pain in my opinion. I saw in your profile you have experience with Swift and I am pretty sure you will love it. 

I will join the Slack group pretty soon. Maybe I can support by writing definition files so at least we would have autocomplete in the IDE and some type checking when using v2.

**Comment by fishcharlie** [comment]
@MickL Just wanna reply to a few points you made.

> It does not add any complexity layer or whatever.

> There are no problems with the compiler output being not what you expected.

<img width="975" alt="image" src="https://user-images.githubusercontent.com/860375/78522001-1aad2080-7789-11ea-9495-6f960551477e.png">
<img width="978" alt="image" src="https://user-images.githubusercontent.com/860375/78522007-1f71d480-7789-11ea-9ec4-3fd3f8d0d211.png">

1. Checkout `master`
2. `npm install`
3. `npm run build`
4. Replace all instances of `lib` to `dist` in the `test` folder. Did this in Atom but any text editor should be able to do this.

Under this scenario all the tests should pass. So running `mocha` should have all the tests be successful, since in theory, the output of TypeScript should be exactly the same. No `.ts` files in the codebase.

But under this scenario we got 8 test failures.

I can understand the test for `utils.empty_function` failing. TypeScript probably messes with the whitespace and the way I wrote that test was **super** janky. Still frustrating, but understandable.

But the other 7 test failures, are unacceptable. All tests succeed when using the `lib` folder as opposed to the `dist` folder. I haven't dug into the tests to figure out why they are failing. Nor will I tonight. This was a quick little project just to start getting my feet wet with it.

But I feel like this is a great example of why TypeScript and build systems are a struggle. There is a **lot** of complexity it adds.

- Lint tools have to be updated (not even sure how this works with ESLint, looks like they support TypeScript linting, but can't find any easy tutorials for how to do this or easily migrate)
- Lot more packages have to be added as dev dependencies
- Tests have to be updated (not even sure how TypeScript works with Mocha and Chai and all that)
- The distribution process needs to be updated (instead of just running `npm publish`, we now have to verify that it builds before and maybe remove the `lib` files, and if there is any error in that process, it can be **REALLY** bad)

There is a **lot** of complexity here. I was honestly hoping that my test would work and all the tests would pass without issue. That way we could start extremely slow and just add the `tsconfig.json` file and not worry about anything I mentioned above. And solve the things above one step at a time. Then after all the things above were solved, we could migrate one file over at a time to TypeScript. But that didn't work.

---

So just wanted to bring this up and kinda go over where I am with TypeScript. It still is on my radar. Progress was made tonight, I got some Atom plugins installed for TypeScript (that I had previously uninstalled). So things are moving in the right direction. But it still remains a low priority because of the things mentioned above.

---

If you would like to try to tackle some of the things I mentioned above I'd be more than happy to take a look at and merge some PRs. My only requirement for that now is that it only adds and prepares for the setup. It shouldn't actually remove or change anything (or any process).

I could also use some support in figuring out some of the things mentioned above. I'm sure overtime I'll spend some time Googling around and researching those items. But learning from others can sometimes be more useful. So if you have any input I'd love to hear it!!

**Comment by MickL** [comment]
I think, as you said, it is not a good idea to move to TS right now. 

First off I dont think you can just move over a few files. The compiler will be full of errors. A common option is noImplicitAny but currenty everything is any. 

For testing I dont think it works like this. Angular uses Jasmine and Karma. Nest uses Jest. They are written in TS and refer to TS code (like Classes etc).

For the project setup there are some starter templates. This is kind of the most complex thing. How does it work with npm? How is it built? How can a non TS project use it? Etc... I would think about that maybe as the last to prevent frustration. 

For eslint there was the great tslint which became deprecated in favor of eslint. There is a nice package typescript-eslint. here you see a very good configuration: https://github.com/nestjs/nest/blob/master/sample/01-cats-app/.eslintrc.js (Angular did a great job in defining good styleguides)

If going for TypeScript I would first start by adding types to all functions and solve all compile errors. Maybe things could be done more easy also? Before that I would not check the tests. 

Anyway for now I would just write definition files so Dynamoose v2 can be used nicely within TS projects. This step is quite fast and easy but when you do it it would be way faster because you know all the functions and types. 

I will join your Slack channel.

**Comment by acerbisgianluca** [comment]
Hi, I got interested in Dynamo a few months ago and I read everywhere that it is better to use a single table design to reduce resource cost.
It is also written in Dynamo's documentation and in almost every "best practice" on the web so I created my first app using DDB directly.
I was wondering if you have noticed some kind of performance degradation using Dynamoose with a table for every model?

**Comment by dolsem** [comment]
@acerbisgianluca Dynamoose was built with the multi-table approach in mind, but it doesn't mean it cannot be used with a single table design. I'm in the process of switching to it myself. What do you mean by "performance degradation"? App-side performance or database performance? Database performance depends heavily on your data models and whether they meet your access patterns well. I don't expect Dynamoose to be slower dealing with multiple tables, in fact, it should be on par, and might in fact be a tad faster.

**Comment by fishcharlie** [comment]
Hi @acerbisgianluca! 👋 

This is a pretty complicated question to answer. The short answer is, it depends on your situation and use case. For example, if you have one document/item in your table, doing a `scan` vs `get` will have such a minor performance difference. Of course as you scale to more documents the difference between `scan` and `get` gets more and more substantial. That is a very specific example (and not *exactly* what you were talking about), but it's important to understand that it really depends on your use case. And the same concepts apply I believe.

I personally haven't done much A/B testing between single table designs vs multi table designs. Although, from my experience, the more important thing to focus on is trying to convert `scan` operations to `query` operations. From what I've seen in DynamoDB architecture design, that is the most common mistake that I see people make, that tends to be an easy performance win with a bit of refactoring. It's a **lot** easier to get people to convert from `scan` to `query`, than to redesign and rethink their entire DB structure. The latter is a huge undertaking. Changing people's mindset around DB design is not an easy task. I'm glad AWS and the community is focusing on it, but despite that, it's not easy to do. So I think little changes adding up overtime is a very positive thing. Keep in mind, not denying about single table design being best practice. Single table design is best practice for DynamoDB in the **vast majority** of cases. But understanding **why** is important (will get into that more later).

Some personal experiences. There are certain situations where I do have multiple tables for one application/service. I think there are use cases for that that make sense. I think the important thing to recognize here, is to try to understand the underlying concepts of DynamoDB. Understanding that `scan` operations look through every document in your table is a very important concept to understand. Understanding that filters occur after reading every document is also important to understand. There is no optimization for filters. Etc. Those are the important concepts to understand. Understanding those concepts are more important than just saying `single table design` in my opinion. Because they help understand **why** single table design is recommended. And when you run into scaling problems, you can understand and debug **why**. It's the why that is important.

But I've never done any A/B testing in terms of table for every model vs single table design.

---

Ok so now that I covered all that. Gonna bring this back around to Dynamoose since this is what this is about.

Highly recommend you read https://github.com/dynamoose/dynamoose/issues/768#issuecomment-608160537 to start (if you haven't already). Those are some future goals for what I wanna do with Dynamoose.

It's **SUPER IMPORTANT** to understand that it's possible to do any of this with Dynamoose currently (even without the future changes in that comment). Dynamoose does not lock you into multi table design patterns. It just does things like allows you to easily customize or modify your data, ensure your data conforms to correct types and schemas, has a much higher level API than `aws-sdk` and is more enjoyable to work with. Which don't care about your DB design pattern. I personally **strongly dislike** working with `aws-sdk` for DynamoDB. Don't get me wrong, it's great for the use case it provides. There are cases where I use it because it's the best tool out there for what I need to do. But for any type of production level application, I **always** use Dynamoose in my projects.

But for example, instead of having nested schemas, currently you can just set it to type `Object` and define the subschema within the existing schema. You can't pass schema's into another schema (YET), but you can get around that by using type `Object`. Instead of having different options for Schema's, you could just define your properties, and by default all properties defined in a Schema are optional. So you could have different objects that are just of type object, and when creating your schema pass them in using the spread operator. It'll be cool and cleaner syntax when you can pass multiple schemas into a model, but it's still possible to implement this design now. The last two items in that chart are a bit more complicated. But totally possible to implement in Dynamoose. Might just require a slight wrapper and you to handle it more within your own code.

But again, Dynamoose can handle **ANY** of this. And if not, let me know, and I'll make it a high priority to get an update out ASAP to support it. But I haven't found a use case where that is the case. The goal of Dynamoose is not to dictate table design. It's to make it easier to work with DynamoDB.

---

I hope this answers your question @acerbisgianluca. Apologizes for the long message.

**Please** don't hesitate to reach out if you have any more questions. You can find more of my [contact information here](https://charlie.fish/contact). There is also a Dynamoose Slack channel, which you can join by clicking the badge in the README of this repo. My name on that Slack is `Charlie Fish`. If you'd like help or support for a specific use case, I'd be happy to chat about it with you here, on Slack, or via one of my other contact methods (in the link above), whichever you prefer.

**Comment by fishcharlie** [comment]
The tl;dr is exactly what @dolsem said:

> but it doesn't mean it cannot be used with a single table design

🎉

**Comment by acerbisgianluca** [comment]
@fishcharlie Thank you for your reply!
I tested Dynamoose a few months ago for my first project based on DynamoDB because I always used Mongoose.
Of course, Dynamoose is a very powerful package for interfacing with Dynamo, but I've never been able to use it with a single table for all the models, so I directly used DDB reading a lot of guides and yesterday I saw a tweet about Dynamoose 2.0 so I came back to see the changes and I will give it another try. Thank you again!

**Comment by fishcharlie** [comment]
@acerbisgianluca Interesting. Would love to hear more about your experiences with single table problems for all your models when using Dynamoose. Maybe there are some more improvements we can make to help meet your needs and use case?

**Comment by acerbisgianluca** [comment]
@fishcharlie When I tried Dynamoose for the first time I hadn't mastered Dynamo yet so I was lacking of knowledge.
Now the Dynamo table is composed by an hash key (called pk), a sort key (called sk) and a few other attributes which are set depending on the "model" of the document. Usually, every pk and sk value is prefixed with the name of the entity plus an '#', something like _USER#218Y4834_, or is a lexicographically sortable string, for example an ISO 8601 date.
Sometimes I also used strings like _LEADERBOARD#it-IT#2020-01_ which are really useful to be used along with _begins_with()_, but I'm sure you know this better than me 😅

**Comment by fishcharlie** [comment]
@acerbisgianluca What about something like the following?

```js
const schema = new dynamoose.Schema({
    "pk": {
        "type": String,
        "hashKey": true // technically not needed since `hashKey` default is the first property in the object, but doing it here for completeness sake
        "set": (val) => `PREFIX#${val}`,
        "get": (val) => val.replace("PREFIX2", "")
    },
    "sk": {
        "type": String,
        "rangeKey": true,
        "set": (val) => `PREFIX2#${val}`,
        "get": (val) => val.replace("PREFIX2", "")
    }
    // ... More attributes here
});
```

Basically that will add the prefix before sending it to DynamoDB, and remove it before getting your document. That way filtering and querying and everything like that, you can use the prefix just fine, but it doesn't interfere with your actual document that you receive back in the responses. You could also just remove the `get` and change the `set` to check to see if the prefix already exists, if not add it. And that way in responses, you can still have the prefix if you want that.

Not sure I'm totally following, but I think that might be what you are looking for. If so, let me know, and we can consider adding a `prefix` property to the attribute settings.

> the common attributes and when I instance a model (new ModelName(...)) I've to add the attributes related to the entity which I'm dealing with (always using the same model with the same schema)

I'm not sure I'm following on that part. Are you looking for [`default`](https://dynamoosejs.com/guide/Schema#default-value--function--async-function)?

Not quite following what exactly you are looking for here.

**Comment by acerbisgianluca** [comment]
> Not quite following what exactly you are looking for here.

Sorry for my bad english, I will try to explain it with an example.

Imagine that I've a table which contains two kinds of entity: User and Game.
I create the following Schema:
```
const schema = new dynamoose.Schema({
    "pk": {
        "type": Object, // Can I do this and get an Object as set argument?
        "hashKey": true
        "set": (val) => `${val.entity}#${val.id}`,
        "get": (val) =>  // split the string on # and return an object
    },
    "sk": {
        "type": Object, // Can I do this and get an Object as set argument?
        "rangeKey": true,
        "set": (val) => `${val.entity}#${val.id}`,
        "get": (val) =>  // split the string on # and return an object
    }
    // ... More attributes here
}, {
    "saveUnknown": true, // now I can use attributes which aren't defined in the schema
    "timestamps": true
});

const Entity = dynamoose.model("MyTable", schema);
const user = new Entity({pk: {entity: 'user', id: 1}, sk{....}, name: 'gianluca'}});
const game = new Entity({pk: {entity: 'game', id: 1}, sk{....}, round: 4}});
```
Maybe instead of using an Object as pk and sk (don't know if it's possibile) I can simply pass an already formatted string (entity#id).
Furthermore it would be nice if I could create two separate schemas for those two entities, but without pk and sk, which are defined in the table schema, and join them. An example:
```
const Entity = dynamoose.model("MyTable", schema);
const user = dynamoose.model({tableSchema: schema, entitySchema: userSchema});
const game = dynamoose.model({tableSchema: schema, entitySchema: gameSchema});
```
or
```
const Entity = dynamoose.model("MyTable", schema);
const user = Entity.join(userSchema);
const game = Entity.join(gameSchema);
```
or again create two different schemas with the same key attributes (pk and sk), but with different values inside setter and getter. Then create two models which should work on the same table (maybe recognizing they have the same key attribute names).

User and Game will be store always in MyTable.

**Comment by fishcharlie** [comment]
@acerbisgianluca Ahh. I think I understand.

## Object Set Argument

> // Can I do this and get an Object as set argument?

100% should be able to do `set` on an `Object` type.

Problem you might run into is I haven't actually tested what happens when you return a different type in the `set` than is defined in the property (ex. you set type to be `Object`, but return a `String` in set). That could cause problems. Because in terms of Dynamoose you want this type to be `Object`, and for DynamoDB, you want the type to be `String`. I *think* right now this will fail or at the very least act really weird.

I think this is something we should fix and improve tho. To me this would be a bug. If you could, try changing the type to be `String` since that is what you want the DynamoDB type to be. You should be able to pass in an object to that, and the `set` would get called and convert it to a `String`. But the type should represent the output of `set` or whatever you want to store in DynamoDB. So I think it'd be more likely to work if you do your exact same code with `"type": String` instead of `Object`. But keep the `set` and `get` and all that the same.

Let me know if that works or doesn't. This is an edge case I haven't thought about at all. We should submit an issue to track this and handle it better.

## Multiple Schema's per Model

In terms of your second part. I think what you really mean is something like:

**IMPORTANT: the code below does not work right now, it's part of the vision I had listed in the comment I linked above**

```js
const GameSchema = new dynamoose.Schema({
    "pk": {
        "type": String,
        "hashKey": true,
        // ...
    },
    "sk": {
        "type": String,
        "rangeKey": true,
        // ...
    },
    "round": String,
    // ... Other game schema attributes
}, {
    "timestamps": true
});

const UserSchema = new dynamoose.Schema({
    "pk": {
        "type": String,
        "hashKey": true,
        // ...
    },
    "sk": {
        "type": String,
        "rangeKey": true,
        // ...
    },
    "name": String,
    // ... Other user schema attributes
}, {
    "timestamps": true
});

// *** This code doesn't work yet, but is part of that comment I linked above. Basically in the future I want to be able to add the ability to have multiple schemas for one model. Then when you create a document it will attempt to conform to one schema or the other, and will only fail if neither schema matches ***
const Entity = dynamoose.model("MyTable", [
    GameSchema,
    UserSchema
]);
// ***********************************************

const anUser = new Entity({pk: {entity: 'user', id: 1}, sk{....}, name: 'gianluca'}});
const aGame = new Entity({pk: {entity: 'game', id: 1}, sk{....}, round: 4}});
```

It would be even cooler to do the following:

```js
const SharedSchema = new dynamoose.Schema({
    "pk": {
        "type": String,
        "hashKey": true,
        // ...
    },
    "sk": {
        "type": String,
        "rangeKey": true,
        // ...
    }
}, {
    "timestamps": true
});

const GameSchema = new dynamoose.Schema(SharedSchema, {
    "round": String,
    // ... Other game schema attributes
});
const UserSchema = new dynamoose.Schema(SharedSchema, {
    "name": String,
    // ... Other user schema attributes
});

// *** This code doesn't work yet, but is part of that comment I linked above. Basically in the future I want to be able to add the ability to have multiple schemas for one model. Then when you create a document it will attempt to conform to one schema or the other, and will only fail if neither schema matches ***
const Entity = dynamoose.model("MyTable", [
    GameSchema,
    UserSchema
]);
// ***********************************************

const anUser = new Entity({pk: {entity: 'user', id: 1}, sk{....}, name: 'gianluca'}});
const aGame = new Entity({pk: {entity: 'game', id: 1}, sk{....}, round: 4}});
```

Keep in mind that syntax right above, I haven't decided if that is the best way to do it. Just ideas about having shared schemas and having schemas be more flexible with one another. Maybe a spread operator would be better? Haven't thought that through enough. Any feedback would be great.

---

Now, you **can** do this right now. Here is what I'd do if I were you:

```js
const GameSchema = {
    "round": String, // By default attributes are optional in Dynamoose (setting required would fail if you do a User here)
    // ... Other game schema attributes
};
const UserSchema = {
    "name": String, // By default attributes are optional in Dynamoose (setting required would fail if you do a Game here)
    // ... Other user schema attributes
};

const SharedSchema = new dynamoose.Schema({
    "pk": {
        "type": String,
        "hashKey": true,
        // ...
    },
    "sk": {
        "type": String,
        "rangeKey": true,
        // ...
    },
    // Other shared attributes between `Game` & `User` (if you have any)
    ...GameSchema,
    ...UserSchema
}, {
    "timestamps": true
});

const Entity = dynamoose.model("MyTable", Schema);
```

Important things to note here:

- If you have an identical property name, you must put it in the `SharedSchema` and shared property names **MUST** have the same type and settings, you can't have the same property name for User & Game and have them be different
- `required` doesn't work for properties that aren't shared between `User` & `Game`
- Might be more, but not that I can think of off the top of my head

---

Long story short. It's better in my opinion than `saveUnknown`. But, it could be even better. I'd like to make more improvements in this area especially since I've gotten more feedback about single table design.

PRs are always welcome to help out on these things. They will happen over time (and soon hopefully). We are close to some major break improvements in this area, and it's high on my priority list. But I do need to re-add some last few features from v1 (plugins, `Model.populate`) before I can start working on this. If anyone wants to submit a PR in the meantime, I'm all for it.

**Comment by acerbisgianluca** [comment]
> Long story short. It's better in my opinion than `saveUnknown`

Yes, I agree.


> Ahh. I think I understand.

And you got the point! Have you a documentation about the code? I would be happy to try to implement what you have written above (the second option):
```
const SharedSchema = new dynamoose.Schema({
    "pk": {
        "type": String,
        "hashKey": true,
        // ...
    },
    "sk": {
        "type": String,
        "rangeKey": true,
        // ...
    }
}, {
    "timestamps": true
});

const GameSchema = new dynamoose.Schema(SharedSchema, {
    "round": String,
    // ... Other game schema attributes
});
const UserSchema = new dynamoose.Schema(SharedSchema, {
    "name": String,
    // ... Other user schema attributes
});

// *** This code doesn't work yet, but is part of that comment I linked above. Basically in the future I want to be able to add the ability to have multiple schemas for one model. Then when you create a document it will attempt to conform to one schema or the other, and will only fail if neither schema matches ***
const Entity = dynamoose.model("MyTable", [
    GameSchema,
    UserSchema
]);
// ***********************************************

const anUser = new Entity({pk: {entity: 'user', id: 1}, sk{....}, name: 'gianluca'}});
const aGame = new Entity({pk: {entity: 'game', id: 1}, sk{....}, round: 4}});
```

**Comment by fishcharlie** [comment]
@acerbisgianluca 

> Have you a documentation about the code?

Do you mean documentation about like how to contribute to the internal code base? [Documentation for using Dynamoose can be found here](https://dynamoosejs.com).

Currently the only documentation we have for the internal code base is the [contributing guidelines](https://github.com/dynamoose/dynamoose/blob/master/CODE_OF_CONDUCT.md).

Some of the code has comments. I need to go back and comment the code better.

If you have questions, reach out and I can try to help.

> I would be happy to try to implement what you have written above

**Awesome!!!!** I really appreciate that. I'd start by reading the [contributing guidelines](https://github.com/dynamoose/dynamoose/blob/master/CODE_OF_CONDUCT.md).

I think the major files you are gonna need to modify are `Schema.ts`, `Model.ts` and `Document.ts`. Specifically for `Document.ts`, there is a function called `Document.objectFromSchema`. Specifically within that there is a function called `checkTypeFunction`. There is also a variable called `schemaAttributes` (which is an array of strings representing the property name of each attribute in the schema). You'd likely need to modify it to be a 2d array maybe (not sure if that is the best solution, but it's the one that comes to mind first)?? And update that function to ensure it conforms to at least one schema. It might be useful to store what schema it conforms to in some internal property that isn't accessible from the user.

As mentioned in the contributing guidelines. The more tests the better!!!

In fact. I'd suggest starting by writing a bunch of tests that will fail. I've found that is the best way to work with Dynamoose personally. Due to the fact that Dynamoose has such great tests, and just because of how the code is setup, I've found that writing tests first is best. If you would like, you could write tests, and submit a PR and I can give you feedback as well before you dive into the code.

If you have **any** questions or need any help, please let me know. I can answer questions or help further.

**Comment by fishcharlie** [comment]
👋! Dynamoose v2.3.0 beta 1 has been released with all of the features I mentioned above. Please keep in mind, it's still in beta form, so please report any bugs or problems you encounter to me. I have tagged all of the relevant PRs with this issue in the description so you can see them all referenced above.

I'd love feedback on this! Positive and negative. If you start using it and find it works for you and meets your needs, let me know. If you use it and run into problems, let me know! Looking for a bunch of feedback here. All of the changes are also in beta so if you have suggestions about how to make it better and have suggestions for breaking changes regarding the functionality added, now is the time to bring up those suggestions.

---

Looking ahead. The only thing on my radar to explore further is converting the concept of Dynamoose model from representing a DynamoDB Table, to representing an entity (ie. User, Product, Order, etc.). That way conceptually it's more clear what is going on, and how to manage that.

I don't foresee that happening in the near future tho, and will likely be part of Dynamoose v3 or beyond since it includes breaking changes and completely changes the definition of a model in Dynamoose.

**Comment by fernandotimm** [comment]
Hi guys,

Regarding the sample using Single Table above:
https://github.com/dynamoose/dynamoose/issues/768#issuecomment-621166383

If I execute a query like the following I get "**Index can't be found for query**"
```
const result = await dynamoose
    .model('MyTable', SharedSchema)
    .query('pk')
    .beginsWith('USER#')
    .exec();
```

Please consider that I'm adding data to that single table this way:

```
...
const profile = await dynamoose.model('MyTable', SharedSchema).create({
      pk: 'USER#myuser',
      sk: 'PROFILE#myuser',
      username: 'myuser',
      name: 'My Name',
      email: 'my@email.com',
    });
...
```

I have installed the version 2.3.0-beta1

Any idea why it the query function does not work?
I want to avoid using scan() as I want to implement the recommendation I've seen on AWS re:invent 2019, the same you mentioned on the beginning of this issue, where they recommend using query().

**Comment by fishcharlie** [comment]
@fernandotimm First of all. **Thank you for helping test this beta!!!** I really appreciate you testing the beta and helping find these problems.

Do me a favor tho. Create a new bug report issue for this. I'm guessing this has to do with the logic we are using for query not working with multiple schemas??? Possibly? I'm not 100% sure on that. But if you could create a new issue with the entire template filled out, that'd be amazing. That way I can see your schema, and have a full reproducible example to test and debug with.

> I want to avoid using scan() as I want to implement the recommendation I've seen on AWS re:invent 2019, the same you mentioned on the beginning of this issue, where they recommend using query().

Totally understand. Looks like for your use case `query` is the way to go for sure.

Create that issue so I have a full example and all of the information I need to test & debug and I'll try to investigate to see what is going on.

**Comment by fernandotimm** [comment]
@fishcharlie sure man! Thank **you** for the quick reply and support on that :) I'll create the issue asap.
